Mathlib/LinearAlgebra/Eigenspace/Basic.lean

```diff
@@ -379,13 +379,73 @@ theorem generalizedEigenspace_le_generalizedEigenspace_finrank [FiniteDimensiona
+  intro x hx
+  simp only [SetLike.mem_coe, mem_generalizedEigenspace] at hx ⊢
+  rw [← LinearMap.comp_apply, ← LinearMap.mul_eq_comp, h.eq, LinearMap.mul_eq_comp,
+    LinearMap.comp_apply, hx, map_zero]
+
+lemma mapsTo_iSup_generalizedEigenspace_of_comm {f g : End R M} (h : Commute f g) (μ : R) :
+    MapsTo g ↑(⨆ k, f.generalizedEigenspace μ k) ↑(⨆ k, f.generalizedEigenspace μ k) := by
+  intro x hx
+  simp only [Submodule.coe_iSup_of_chain, mem_iUnion, SetLike.mem_coe] at hx
+  obtain ⟨k, hk : x ∈ f.generalizedEigenspace μ k⟩ := hx
+  simp only [Submodule.coe_iSup_of_chain, mem_iUnion, SetLike.mem_coe]
+  exact ⟨k, f.mapsTo_generalizedEigenspace_of_comm h μ k hk⟩
+
+lemma disjoint_generalizedEigenspace [NoZeroSMulDivisors R M] [IsReduced R]
+    (f : End R M) {μ₁ μ₂ : R} (hμ : μ₁ ≠ μ₂) (k l : ℕ) :
+    Disjoint (f.generalizedEigenspace μ₁ k) (f.generalizedEigenspace μ₂ l) := by
+  rw [disjoint_iff]
+  set p := f.generalizedEigenspace μ₁ k ⊓ f.generalizedEigenspace μ₂ l
+  by_contra hp
+  replace hp : Nontrivial p := Submodule.nontrivial_iff_ne_bot.mpr hp
+  have hc₁ : Commute f (f - algebraMap R (End R M) μ₁) :=
+    Commute.sub_right rfl (Algebra.commute_algebraMap_right μ₁ f)
+  have hc₂ : Commute f (f - algebraMap R (End R M) μ₂) :=
+    Commute.sub_right rfl (Algebra.commute_algebraMap_right μ₂ f)
+  let f₁ : End R p := LinearMap.restrict (f - algebraMap R (End R M) μ₁) (MapsTo.inter_inter
+    (mapsTo_generalizedEigenspace_of_comm hc₁ μ₁ k) (mapsTo_generalizedEigenspace_of_comm hc₁ μ₂ l))
+  let f₂ : End R p := LinearMap.restrict (f - algebraMap R (End R M) μ₂) (MapsTo.inter_inter
+    (mapsTo_generalizedEigenspace_of_comm hc₂ μ₁ k) (mapsTo_generalizedEigenspace_of_comm hc₂ μ₂ l))
+  have hn₁ : IsNilpotent f₁ := by
+    use k; ext ⟨x, hx⟩; simpa [LinearMap.restrict_apply, LinearMap.pow_restrict _] using hx.1
+  have hn₂ : IsNilpotent f₂ := by
+    use l; ext ⟨x, hx⟩; simpa [LinearMap.restrict_apply, LinearMap.pow_restrict _] using hx.2
+  have : IsNilpotent (f₂ - f₁) := by
+    apply Commute.isNilpotent_sub _ hn₂ hn₁
+    ext; simp [smul_sub, sub_sub, smul_comm μ₁, add_sub_left_comm]
+  have hf₁₂ : f₂ - f₁ = algebraMap R (End R p) (μ₁ - μ₂) := by ext; simp [sub_smul]
+  rw [hf₁₂, IsNilpotent.map_iff (NoZeroSMulDivisors.algebraMap_injective _ _),
+    isNilpotent_iff_eq_zero, sub_eq_zero] at this
+  contradiction
+
+lemma disjoint_iSup_generalizedEigenspace [NoZeroSMulDivisors R M] [IsReduced R]
+    (f : End R M) {μ₁ μ₂ : R} (hμ : μ₁ ≠ μ₂) :
+    Disjoint (⨆ k, f.generalizedEigenspace μ₁ k) (⨆ k, f.generalizedEigenspace μ₂ k) := by
+  simp_rw [(f.generalizedEigenspace μ₁).mono.directed_le.disjoint_iSup_left,
+    (f.generalizedEigenspace μ₂).mono.directed_le.disjoint_iSup_right]
+  exact disjoint_generalizedEigenspace f hμ
+
+theorem independent_generalizedEigenspace [NoZeroSMulDivisors R M] [IsReduced R] (f : End R M) :
+    CompleteLattice.Independent (fun μ ↦ ⨆ k, f.generalizedEigenspace μ k) := by
```

Comment:
Instead of showing `Independent` and `disjoint`, can you go straight to showing `DirectSum.Decomposition`? That gives you the other results for free, IIRC.

Reply:
I do plan to use the direct sum API once I've established that I have a family of independent, spanning submodules but I think that has to come later.

However you're right that it's silly to prove pairwise disjointness and then independence (in view of `Independent.pairwiseDisjoint`). Unless I need pairwise disjointness to prove independence I will be removing it.

Anyway thanks for all the remarks, especially given this is still WIP. I will address your other remarks (and split this into several PRs) once I've closed the final sorry.

Reply:
It's easy enough to write down an inverse (indeed this is implicit in my proof) but I think I've taken the easiest route. Note that in general we're proving independence for any endomorphism, not just triangularisable ones, so if I wanted to sue the `DirectSum` machinery here I'd have to be talking about a direct sum of `⨆ μ, ⨆ k, f.generalizedEigenspace μ k` (and I'd have to restrict `f` to this) rather than `M`.